Uses of `pytest`'s `tmpdir` fixture should be replaced with `tmp_path`

### Description

Some of the tests involve writing data, and such data should not be stored permanently. `pytest` has two fixtures that provide temporary directories for writing test data: the recommended [`tmp_path`](https://docs.pytest.org/en/stable/how-to/tmp_path.html#the-tmp-path-fixture), which uses [`pathlib.Path`](https://docs.python.org/3/library/pathlib.html) from the Python standard library, and the legacy [`tmpdir`](https://docs.pytest.org/en/stable/how-to/tmp_path.html#the-tmpdir-and-tmpdir-factory-fixtures) utilizing [`py.path`](https://py.readthedocs.io/en/latest/path.html), a third party library. The following shell command reveals that many tests in `astropy` still use `tmpdir`:
```shell
git grep --files-with-matches tmpdir -- *tests*
```
All uses of `tmpdir` should be replaced with `tmp_path`, although not all sub-packages have to be updated in a single pull request.

### Sub-packages to update

- [x] `config` #13788
- [x] `coordinates` #13797
- [x] `cosmology` #13847
- [x] `io.ascii` #13876
- [x] `io.fits` https://github.com/astropy/astropy/pull/13879
- [x] `io.misc` (but note that `io.misc.asdf` is deprecated so only `io.misc.tests` is worth updating) #13878
- [x] `io.registry` #13849
- [x] `io.votable` #13850
- [x] `modeling` #13799
- [x] `nddata` #13807
- [x] `samp` #13848
- [x] `table` #13802
- [x] `tests` #13800
- [x] `time` #13801
- [x] `utils` #13874
- [x] `visualization` https://github.com/astropy/astropy/pull/13822
- [x] `wcs` #13796

In `astropy/cosmology/tests/test_units.py` under the `test_equivalencies_asdf()` function, the test fails when using `tmp_path` parameter:

```powershell
================================================================================================================================= FAILURES ==================================================================================================================================
_____________________________________________________________________________________________________________________ test_equivalencies_asdf[with_H0] ______________________________________________________________________________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-jerobado/pytest-185/test_equivalencies_asdf_with_H0'), equiv = <function with_H0 at 0x7f3c2d0e2ee0>, recwarn = WarningsRecorder(record=True)

    @pytest.mark.skipif(not HAS_ASDF, reason="requires ASDF")
    @pytest.mark.parametrize("equiv", [cu.with_H0])
    def test_equivalencies_asdf(tmp_path, equiv, recwarn):
        from asdf.tests import helpers

        tree = {"equiv": equiv()}
>       helpers.assert_roundtrip_tree(tree, tmp_path)

astropy/cosmology/tests/test_units.py:400:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/astropy-contributor/lib/python3.9/site-packages/asdf/tests/helpers.py:172: in assert_roundtrip_tree
    _assert_roundtrip_tree(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

tree = {'equiv': [(Unit("1.47798 littleh"), None)]}, tmpdir = PosixPath('/tmp/pytest-of-jerobado/pytest-185/test_equivalencies_asdf_with_H0')

    def _assert_roundtrip_tree(
        tree,
        tmpdir,
        *,
        asdf_check_func=None,
        raw_yaml_check_func=None,
        write_options={},
        init_options={},
        extensions=None,
        tree_match_func="assert_equal",
    ):

>       fname = str(tmpdir.join("test.asdf")) --> PosixPath object no longer valid here
E       AttributeError: 'PosixPath' object has no attribute 'join'

../../miniconda3/envs/astropy-contributor/lib/python3.9/site-packages/asdf/tests/helpers.py:187: AttributeError
========================================================================================================================== short test summary info ==========================================================================================================================
FAILED astropy/cosmology/tests/test_units.py::test_equivalencies_asdf[with_H0] - AttributeError: 'PosixPath' object has no attribute 'join'
======================================================================================================================= 1 failed, 24 passed in 0.62s ========================================================================================================================
```
This line `fname = str(tmpdir.join("test.asdf"))` raises an error because the argument **tmpdir** is now a pathlib.Path object (PosixPath).

The original type/value of **tmpdir** was:
```Python
tmpdir=local('/tmp/pytest-of-jerobado/pytest-188/test_equivalencies_asdf_with_H0')
```

**Questions**

1. Is `../../miniconda3/envs/astropy-contributor/lib/python3.9/site-packages/asdf/tests/helpers.py:187` a 3rd party library? Are we allowed to make changes here?

2. Should I keep the parameter as is in `test_equivalencies_asdf()` for the meantime or do you have any alternative without touching the 3rd party file?

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -339,7 +339,7 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
             # text mode, which won't work. .read_binary() does, and
             # surely other ducks would return binary contents when
             # called like this.
-            # py.path.LocalPath is what comes from the tmpdir fixture
+            # py.path.LocalPath is what comes from the legacy tmpdir fixture
             # in pytest.
             fileobj = io.BytesIO(fileobj.read_binary())
         except AttributeError:

--- a/astropy/utils/tests/test_data.py
+++ b/astropy/utils/tests/test_data.py
@@ -146,7 +146,7 @@ def readonly_dir(d):
 @pytest.fixture
 def readonly_cache(tmp_path, valid_urls):
     with TemporaryDirectory(dir=tmp_path) as d:
-        # other fixtures use the same tmpdir so we need a subdirectory
+        # other fixtures use the same tmp_path so we need a subdirectory
         # to make into the cache
         d = pathlib.Path(d)
         with paths.set_temp_cache(d):
@@ -179,7 +179,7 @@ def fake_readonly_cache(tmp_path, valid_urls, monkeypatch):
                       "_SafeTemporaryDirectory monkeypatched out")
 
     with TemporaryDirectory(dir=tmp_path) as d:
-        # other fixtures use the same tmpdir so we need a subdirectory
+        # other fixtures use the same tmp_path so we need a subdirectory
         # to make into the cache
         d = pathlib.Path(d)
         with paths.set_temp_cache(d):
@@ -627,7 +627,7 @@ def test_download_certificate_verification_failed():
 
     with pytest.warns(AstropyWarning, match=msg) as warning_lines:
         fnout = download_file(TESTURL_SSL, cache=False,
-                ssl_context=ssl_context, allow_insecure=True)
+                              ssl_context=ssl_context, allow_insecure=True)
 
     assert len(warning_lines) == 1
     assert os.path.isfile(fnout)
@@ -667,7 +667,7 @@ def test_download_parallel_from_internet_works(temp_cache):
 @pytest.mark.parametrize("method", [None, "spawn"])
 def test_download_parallel_fills_cache(tmp_path, valid_urls, method):
     urls = []
-    # tmpdir is shared between many tests, and that can cause weird
+    # tmp_path is shared between many tests, and that can cause weird
     # interactions if we set the temporary cache too directly
     with paths.set_temp_cache(tmp_path):
         for um, c in islice(valid_urls, FEW):

--- a/astropy/visualization/wcsaxes/tests/test_images.py
+++ b/astropy/visualization/wcsaxes/tests/test_images.py
@@ -48,8 +48,8 @@ class TestBasic(BaseImageTests):
     def test_tight_layout(self):
         # Check that tight_layout works on a WCSAxes.
         fig = plt.figure(figsize=(8, 6))
-        axs = [fig.add_subplot(2, 1, i, projection=WCS(self.msx_header))
-               for i in [1, 2]]
+        for i in (1, 2):
+            fig.add_subplot(2, 1, i, projection=WCS(self.msx_header))
         fig.tight_layout()
         return fig
 
@@ -659,9 +659,10 @@ class TestBasic(BaseImageTests):
 
         with pytest.warns(AstropyUserWarning,
                           match="Received `center` of representation type "
-                                "<class 'astropy.coordinates.representation.CartesianRepresentation'> "
+                                "<class 'astropy.coordinates.representation.CartesianRepresentation'> "  # noqa: E501
                                 "will be converted to SphericalRepresentation"):
-            r3 = SphericalCircle(SkyCoord(x=-0.05486461, y=-0.87204803, z=-0.48633538, representation_type='cartesian'),
+            r3 = SphericalCircle(SkyCoord(x=-0.05486461, y=-0.87204803, z=-0.48633538,
+                                          representation_type='cartesian'),
                                  0.15 * u.degree, edgecolor='purple',
                                  facecolor='none', transform=ax.get_transform('fk5'))
 
@@ -693,8 +694,8 @@ class TestBasic(BaseImageTests):
 
         # Add a rectangle patch (100 degrees by 20 degrees)
         r = Rectangle((255, -70), 100, 20,
-                       label='Rectangle', edgecolor='red', facecolor='none', linestyle='--',
-                       transform=ax.get_transform('icrs'))
+                      label='Rectangle', edgecolor='red', facecolor='none', linestyle='--',
+                      transform=ax.get_transform('icrs'))
         ax.add_patch(r)
 
         ax.coords[0].set_ticklabel_visible(False)
@@ -703,7 +704,7 @@ class TestBasic(BaseImageTests):
         return fig
 
     @figure_test
-    def test_beam_shape_from_args(self, tmpdir):
+    def test_beam_shape_from_args(self, tmp_path):
         # Test for adding the beam shape with the beam parameters as arguments
         wcs = WCS(self.msx_header)
         fig = plt.figure(figsize=(4, 3))
@@ -718,7 +719,7 @@ class TestBasic(BaseImageTests):
         return fig
 
     @figure_test
-    def test_beam_shape_from_header(self, tmpdir):
+    def test_beam_shape_from_header(self, tmp_path):
         # Test for adding the beam shape with the beam parameters from a header
         hdr = self.msx_header
         hdr['BMAJ'] = (2 * u.arcmin).to(u.degree).value
@@ -736,7 +737,7 @@ class TestBasic(BaseImageTests):
         return fig
 
     @figure_test
-    def test_scalebar(self, tmpdir):
+    def test_scalebar(self, tmp_path):
         # Test for adding a scale bar
         wcs = WCS(self.msx_header)
         fig = plt.figure(figsize=(4, 3))

--- a/docs/conftest.py
+++ b/docs/conftest.py
@@ -27,15 +27,18 @@ os.mkdir(os.path.join(os.environ['XDG_CACHE_HOME'], 'astropy'))
 
 @pytest.fixture(autouse=True)
 def _docdir(request):
-    """Run doctests in isolated tmpdir so outputs do not end up in repo"""
+    """Run doctests in isolated tmp_path so outputs do not end up in repo"""
     # Trigger ONLY for doctestplus
     doctest_plugin = request.config.pluginmanager.getplugin("doctestplus")
     if isinstance(request.node.parent, doctest_plugin._doctest_textfile_item_cls):
-        # Don't apply this fixture to io.rst.  It reads files and doesn't write
+        # Don't apply this fixture to io.rst.  It reads files and doesn't write.
+        # Implementation from https://github.com/pytest-dev/pytest/discussions/10437
         if "io.rst" not in request.node.name:
-            tmpdir = request.getfixturevalue('tmpdir')
-            with tmpdir.as_cwd():
-                yield
+            old_cwd = os.getcwd()
+            tmp_path = request.getfixturevalue('tmp_path')
+            os.chdir(tmp_path)
+            yield
+            os.chdir(old_cwd)
         else:
             yield
     else:

